py-petsc4py fails on all systems, all arch

`py-petsc4py@3.14.0` build fails using `spack@develop` (7dcf3f7aedbfdf085451f8fffb2ecdfb08c96c7d from `Tue Feb 2 11:11:51 2021 -0800`)
* `Ubuntu 18.04` w/ `gcc@7.5.0`
* `Ubuntu 20.04` w/ `gcc@9.3.0`
* `RHEL 7` w/ `gcc@8.1.0`
* `RHEL 8` w/ `gcc@8.3.1`

Concrete spec: [py-petsc4py-endo3v.spec.yaml.txt](https://github.com/spack/spack/files/5914488/py-petsc4py-endo3v.spec.yaml.txt)
Full build log: [py-petsc4py-build-out.txt](https://github.com/spack/spack/files/5914490/py-petsc4py-build-out.txt)
Concretize output: [py-petsc4py-dag.txt](https://github.com/spack/spack/files/5914497/py-petsc4py-dag.txt)

```
$> spack install -f ./py-petsc4py-endo3v.spec.yml
...
==> Installing py-petsc4py-3.14.0-endo3vxvolun3ajpssfvyqgypqzkrh43
==> Using cached archive: /opt/spack/var/spack/cache/_source-cache/archive/33/33ac9fb55a541e4c1deabd6e2144da96d5ae70e70c830a55de558000cf3f0ec5.tar.gz
==> py-petsc4py: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/python-3.8.6-4aev2ilejphuclznv7zazjxohanwz46i/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'

2 errors found in build log:
     42    src/petsc4py.PETSc.c:326298:30: note: variable tracking size limit exceeded with '-fvar-tracking-assignments', retrying without
     43    326298 | static CYTHON_SMALL_CODE int __pyx_pymod_exec_PETSc(PyObject *__pyx_pyinit_module)
     44           |                              ^~~~~~~~~~~~~~~~~~~~~~
     45    /opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.3.2-knmzscajrauwght5atlc2vevwjzwn2ke/bin/mpicc -pthread -fPIC -fPIC -Wno-unused-result -Wsign-compare
           -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/zlib-1.2.11-p7dmb5pqi6svirhovljrkp3uyfiytz3m/include -I/opt/spack/opt/sp
           ack/linux-ubuntu20.04-x86_64/gcc-9.3.0/metis-5.1.0-lxkd2thzi5xc5z3aqvw4eodot57r53y5/include -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/parmetis-4.0.3-7
           b3zkab7n54xkxkba2etquagx46mr6ze/include -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/hdf5-1.10.7-t4jqnkp2mejfioelbtantmf7t2vdnoup/include -I/opt/spack/op
           t/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/superlu-dist-6.4.0-qn52fuce45s64duiwjrv3n5nugai6f2p/include -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/hypre
           -2.20.0-celzcb34v2s6hppgktqyrdpubyp4rk5g/include -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/petsc-3.14.1-wakbzwrwms2g3xcyc643cnof4s4lf23r/include -Isrc
           /include -I/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/py-numpy-1.20.0-q5r54n5m4zhq6ujwxaaqqlm2hovb54lv/lib/python3.8/site-packages/numpy/core/include -I/
           opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/python-3.8.6-4aev2ilejphuclznv7zazjxohanwz46i/include/python3.8 -c src/libpetsc4py.c -o build/temp.linux-x86_64
           -3.8/src/libpetsc4py.o
     46    /opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.3.2-knmzscajrauwght5atlc2vevwjzwn2ke/bin/mpicc -pthread -fPIC /opt/spack/lib/spack/env/gcc/gcc -shared
            -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/bzip2-1.0.8-5m22qwv5akdogbbq3qbuhilvi
           ppzqeli/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/expat-2.2.10-r4dctmf674e37q4jc2pvphiy5ik4riz3/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_6
           4/gcc-9.3.0/gdbm-1.18.1-iowiob4u4kujr6pyhbolir3nokuqylwh/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/gettext-0.21-yt6n2phbjvscurpwbkm2gfx52rsd4rkm/l
           ib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/libffi-3.3-hvrhttuiahh5krx6w4oskf3mryadojct/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0
           /ncurses-6.2-pzjdrlluvcj4s6ah6lxbd77vok3eoeaz/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/openssl-1.1.1i-6peexc2wtn22oxuevnyhrimerrpmqxeh/lib -L/opt
           /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/readline-8.0-do42wztg5acct7tiuewpgnvgktwrkudh/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/sqlite
           -3.33.0-2vrxq4lf2wh2qac4v2ngc2eqbna47oej/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/util-linux-uuid-2.36-qm5gupet6fiuoat4m2rnlusdwb5xiot7/lib -L/op
           t/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/xz-5.2.5-kamfyq4jqy3je7qoimrp2tgs6gqswzcf/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/zlib-1.2.
           11-p7dmb5pqi6svirhovljrkp3uyfiytz3m/lib build/temp.linux-x86_64-3.8/src/PETSc.o build/temp.linux-x86_64-3.8/src/libpetsc4py.o -L/opt/spack/opt/spack/linux-ubuntu20.04
           -x86_64/gcc-9.3.0/petsc-3.14.1-wakbzwrwms2g3xcyc643cnof4s4lf23r/lib -L/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/python-3.8.6-4aev2ilejphuclznv7zazjxohan
           wz46i/lib -Wl,--enable-new-dtags,-R/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/petsc-3.14.1-wakbzwrwms2g3xcyc643cnof4s4lf23r/lib -Wl,--enable-new-dtags,-R
           /opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/python-3.8.6-4aev2ilejphuclznv7zazjxohanwz46i/lib -lpetsc -o build/lib.linux-x86_64-3.8/petsc4py/lib/PETSc.cpy
           thon-38-x86_64-linux-gnu.so
     47    /opt/spack/lib/spack/env/gcc/gcc: file not recognized: file format not recognized
  >> 48    collect2: error: ld returned 1 exit status
  >> 49    error: command '/opt/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.3.2-knmzscajrauwght5atlc2vevwjzwn2ke/bin/mpicc' failed with exit status 1
```

This breakage happened fairly recently (~ last month). I will try a bisect to see when this broke.

@balay @adamjstewart

--- a/var/spack/repos/builtin/packages/py-petsc4py/ldshared.patch
+++ b/var/spack/repos/builtin/packages/py-petsc4py/ldshared.patch
@@ -1,0 +1,13 @@
+diff --git a/conf/baseconf.py b/conf/baseconf.py
+index 8a2466a5bd..73c08b923a 100644
+--- a/conf/baseconf.py
++++ b/conf/baseconf.py
+@@ -211,7 +211,7 @@ class PetscConfig:
+         ldshared = getenv('LDSHARED', ldshared)
+         ldflags = getenv('LDFLAGS', cflags + ' ' + (ldflags or ''))
+         ldcmd = split_quoted(ld) + split_quoted(ldflags)
+-        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd]
++        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd and (flg.find('/lib/spack/env')<0)]
+         ldshared = str.join(' ', ldshared)
+         #
+         def get_flags(cmd):

--- a/var/spack/repos/builtin/packages/py-petsc4py/package.py
+++ b/var/spack/repos/builtin/packages/py-petsc4py/package.py
@@ -31,6 +31,8 @@ class PyPetsc4py(PythonPackage):
 
     variant('mpi', default=True,  description='Activates MPI support')
 
+    patch('ldshared.patch')
+
     depends_on('py-cython', type='build', when='@develop')
     depends_on('python@2.6:2.8,3.3:', type=('build', 'run'))
     depends_on('py-setuptools', type='build')

